### DISCUSSION TITLE
Name and location of the goal, gosl, tensor, plot "data science and visualization" framework

### DISCUSSION THREAD

**Discussion by rcoreilly** [cogentcore/lab#11] [Ideas]
After much work, the goal, gosl, tensor and plot packages are now relatively stable and working well for the emer neural network simulations. Now we need to decide how to package all of this stuff.

* Dependency-wise, all of these packages depend on each-other, and on lots of other things within core, but nothing else in core depends on them.

* From a basic semantics and categorization of software functionality perspective, it doesn't really make sense to have a "data science and visualization" system embedded within a GUI framework.

* The external visibility of the framework as a coherent entity unto itself is also enhanced by it being a separate entity.
 
* Also, logistically, do we really want people filing issues and having discussions about math and algorithms and all that stuff in the same space as low-level GUI stuff? No.

Therefore, it makes sense to move this stuff out of core, and into its own repository.

Although we could put it into the `cogent` monorepo, I think it is sufficiently complex and multi-faceted as to warrant its own separate repository under `cogentcore.org`.  The `cogent` monorepo is more appropriate for self-contained apps, but not for something that is really a new toolkit and framework.

To make this all happen, we need an overall _name_ for the thing.

In terms of the overall scope, this framework encompasses the functionality of Python's `NumPy` and associated packages like `matplotlib`, and the broader [IPython](https://ipython.org/) ecosystem ([Jupyter](https://jupyter.org/) notebooks etc). A central focus is on visualization and interactive GUI elements that make the exploration and management of data easier, and a "capstone" package in the system is a core data <-> visualization API that integrates everything and makes it easy to generate any kind of visualization from any kind of data, and have that connection be bidirectional (and yet clean, simple, etc).

By leveraging the strengths of the Cogent Core GUI framework, we can potentially make the system "even better" than the current python offerings, and there is no doubt that they will be much cleaner at the code level (being in just one language instead of a mess of C++ and python) and automatically available on the web without requiring a backend server, etc.

These wikipedia pages provide lots of references for the overall scope and content: https://en.wikipedia.org/wiki/Data_and_information_visualization https://en.wikipedia.org/wiki/Category:Free_data_visualization_software https://en.wikipedia.org/wiki/Category:Data_visualization_software https://en.wikipedia.org/wiki/List_of_information_graphics_software

Also, here are our existing discussions for the design of the system, which ended up in the cogent repo:
* https://github.com/cogentcore/cogent/discussions/324
* https://github.com/cogentcore/cogent/discussions/348

Some broad considerations on naming:
* We will retain the "Cogent" prefix, and the `cogent` repo apps have established the "generic" theme where the mail program is just `Mail` (i.e., "Cogent Mail"), instead of trying to come up with some kind of unique name for the app itself (e.g., "thunderbird" or "outlook").
* OTOH, we didn't call "core" "Cogent GUI", in part because it is _bigger_ than that, and the same consideration applies to this project, because it is an entire _framework_, not just an encapsulated well-defined app. This is the same reasoning for why it doesn't belong under the existing `cogent` monorepo.
* Strongly prefer single, short, "real" words, as contrasted with acronyms and contractions (bonus points for cool backronyms as in core = code once run everywhere, but it also has to stand on its own).

Here's an initial list of names:

* `Numbers`: this is what we originally came up with in the generic, self-contained app mode.
* `Data`: another appropriate generic name.

Neither of those capture the key _visualization_ aspect of the framework, however. In my opinion, some connection to this term is essential to capture the essence of the project.

* `Vision`: this is my favorite, because it is the shortest coherent "real" word that is closely related to "Visualization", and it has clear connotations of "insight" and is used for things like a [Vision Statement](https://en.wikipedia.org/wiki/Vision_statement) -- we want people to "have a vision" of their data -- not just to see it, but to really understand it deeply, and this term captures that "deep understanding" aspect.
* `Insight`: per above, but less cool than Vision IMO. OTOH, @kkoreilly thinks "vision" is too confusable with _actual_ vision stuff like the Apple Vision Pro and some kind of other vision software, like an AI vision model or something. I'm not convinced that this is a problem.
* `Sight`: another shorter option along the same lines. "Cogent Sight" just doesn't quite roll off the tongue as nicely as "Cogent Vision" however.

Some other more jokey options:

* `Ize`: as in visualize, and it sounds like "eyes".
* `Lize`: statistics are just lies.. :)

**Comment by kkoreilly** [comment]
We just recently had a conversation in which you referenced https://github.com/emer/vision, and I had to clarify what you were talking about because of your confusing Cogent Vision idea. Cogent Insight makes more sense than Cogent Vision, but it seems strange; almost satirical.

Moreover, your point here indirectly raises a key issue that I have mentioned before:

> Also, logistically, do we really want people filing issues and having discussions about math and algorithms and all that stuff in the same space as low-level GUI stuff? No.

If you take the approach of cogentcore/core#1380 (some interrepo mirroring system), then people will randomly file issues in either one of the repositories, leading to a massive headache. If you disable issues on the smaller repo, then there is really no point in having it, and you get the exact same problem as golang.org/x, where people have to file issues on the main repo. It would be even worse than that since you wouldn't be able to have PRs in the smaller repo or else you would get massive synchronization problems. Your cogentcore/core#1380 proposal is simply untenable, as it sacrifices a massive amount of clarity and usability in exchange for a minor improvement in discoverability.

As such, you need to clarify whether you are proposing taking a cogentcore/core#1380-esque approach here. If so, that is not happening. If not, then I think it is a highly questionable choice to remove key functionality that strongly depends on core from the repo. We want an integrated documentation content wiki for this ecosystem, and as soon as you poke holes in the monorepo, you get a lot of problems.

**Reply by rcoreilly** [reply]
Definitely would _not_ want to do a mirroring thing with this `Vision` repo.

> If not, then I think it is a highly questionable choice to remove key functionality that strongly depends on core from the repo. We want an integrated documentation content wiki for this ecosystem, and as soon as you poke holes in the monorepo, you get a lot of problems.

You need to be more specific here. Depending on core is not at all a problem. The key point is that the rest of core absolutely does not depend on `Vision` (if I use this enough times, you'll come to love it ;). 

The documentation for `Vision` will have exactly the same _asymmetric_ (one way) connectivity structure as the dependency graph, and hopefully the intra-content linking system can make this nice and easy: `Vision` will refer to `core` but I don't see any reason why `core` would need to refer to `Vision`?

Exactly what "problems" are you envisioning?

And I will be more than happy to rename https://github.com/emer/vision to make way for `Cogent Vision` :)

**Reply by kkoreilly** [reply]
The whole point is that vision is a perfect name for emer/vision because that it is precisely what it does; vision does not make any sense for this.

The issue with Cogent Data depending heavily on core is that changes to core will frequently require changes to Cogent Data, so we will constantly have to update dependencies and file multiple PRs for every change; the advantage of the monorepo is you avoid that.

As I have articulated previously, intrawiki linking works a lot better than interwiki linking in terms of getting the formatted name and metadata. Moreover, core will need to refer to Data quite frequently, as it provides alternatives to widgets like tables. Also, data science is one of the key uses of Cogent Core, and having plotting tutorials and examples alongside the other tutorials is a major advantage.

**Reply by rcoreilly** [reply]
This dependency on core also holds with the `cogent` repo.

And the converse point also holds in both of these cases: you _want_ the higher-level packages to be more stable and not forcing them to update immediately to every little change in `core`.

I have articulated exactly why `Vision` _does_ make sense.  Sure, there is no doubt that it _also_ makes sense for an application that literally deals with vision, but that does _not_ invalidate the points raised about why it is a great name for a data visualization and analysis framework.

I have also clearly articulated why simple generic names like `Data` and `Numbers` are inadequate. It is incumbent upon you to come up with another alternative that is better than `Vision` at this point. And continuing to press the point about it _also_ being relevant for some other more narrow vision-specific app is _not_ a sufficient basis for rejecting it for this _much more important_ case, IMO.

**Reply by rcoreilly** [reply]
> Moreover, core will need to refer to Data quite frequently, as it provides alternatives to widgets like tables. Also, data science is one of the key uses of Cogent Core, and having plotting tutorials and examples alongside the other tutorials is a major advantage.

It is unproblematic for core to have a generic pointer to `Vision` for anyone interested in applications of the Core technology to the domain of data science and visualization. But again, these are precisely that: _applications_ of the core technology. They are _not_ the core technology itself, and they do _not_ belong in the same repository as the core GUI infrastructure.

**Reply by rcoreilly** [reply]
@kkoreilly objecting to the name `Vision` according to your argument is like saying that Apple is a terrible name for a computer company because they don't sell apples.

**Reply by rcoreilly** [reply]
The value of a name is not in its _discoverability_ in some bizarre scenario where you only see the name and no other context as to what it does. Rather, it is in how it makes the users _feel_ when they see it, both early on and later after actually using it, about how "cool" and "apt" of a name it is for this thing that sounds _interesting_ and _powerful_ and _cool_.

**Reply by kkoreilly** [reply]
Okay, based on your logic should we make the following renames?

* Cogent Mail -> Cogent Connection
* Cogent Code -> Cogent Creation
* Cogent Terminal -> Cogent Control
* Cogent Canvas -> Cogent Cool
* Cogent Web -> Cogent Discovery
* Cogent Files -> Cogent Information

Once you do that, you have a lot of very vague names that may connote "power" and "interesting" and "cool", but if people can't remember which is which, the value is significantly diminished.

It is not about "discoverability" per se but about *memorability*: if there are 30 Cogent somethings, you want to be able to say, oh Cogent Mail is the one for mail and Cogent Data is the one for data science, instead of conflating Cogent Connection, Cogent Vision, and Cogent Discovery in your memory. I think that Cogent Data and Cogent Science are the two best options so far; we are making a data science framework, not a blog article peppered with buzzwords for SEO optimization on business and technology trends.

As we have learned time and time again from Go and Cogent Core, programming isn't about creating an intricate masterpiece but rather a clear and elegant framework that works without bloat. You don't need to oversell this to people; `numpy` and `matlab` don't conjure some grand vision of philosophical epiphany, but rather they just tell you what they do, and people seem to like that. Cogent is plenty of branding, and Data or Science complements that with reality.

If Cogent Science isn't a sufficiently "cool" name for users, then they are probably not the intended audience. It is somewhat ironic that the person who came up with the names "goki", "grease", "gide", "gopi", "PerTrlMSec", "TstTrlPlt", etc is now concerned about the "coolness" of names.

**Reply by rcoreilly** [reply]
Key points (made already above): this is one of _3_ top-level repositories under Cogent, not one of a large number of apps with well-established category names that everyone knows and recognized.

There isn't any accepted standard name for this category of 'framework for data science and visualization`.  If there was, we might well use it. But there isn't. So all of your silly names above are moot. Please stay focused on the issue at hand. Data is not a sufficient name. Find another or accept that Vision or Vivid or some other suitable evocative name is better.

**Reply by rcoreilly** [reply]
I like Cogent Science well enough, and the term means "the state of knowing" which certainly captures the "deep knowledge" aspect that I also think Vision captures well enough, and that Data definitely does not capture. Science doesn't directly capture "visualization" but I guess we can't have everything in one word.

SciPy is also a precedent for Science and "scientific computing": https://scipy.org/

**Reply by kkoreilly** [reply]
As the title of this discussion states, this is a data science and visualization framework built on Cogent Core. As such, the title can really only use these words:

* Cogent
* Data
* Science
* Visualization

Cogent Visualization doesn't really work, so the two options are Cogent Data and Cogent Science as I stated above. I am not sure Cogent Science makes sense; like Cogent Vision, it seems oversold, and Data is a more accurate descriptor. However, Cogent Science is certainly better than Cogent Vision, and so it is a potential compromise.

**Reply by rcoreilly** [reply]
FWIW, https://en.wikipedia.org/wiki/Category:Data_visualization_software has the following "Data" names: DataScene, DataViva, and Datawatch Corporation, which all combine Data with some kind of visualization modifier.  There are various "Viz" and "Vis" examples there, and in https://en.wikipedia.org/wiki/Category:Free_data_visualization_software there are 4 Vis* examples (and no other Data* examples).  Chart and Plot names are overall the most common probably.

Also, the single closest example to what we are trying to achieve is called "Jupyter" so whatever. If you search for alternatives to that, you get various things including Kaggle, Google Colab, Deepnote, SageMaker, Apache Zeppelin, Databricks, Datalore, RStudio, Observable, DataCamp Light.  

So there isn't much of an obvious pattern here, with plenty of random names, but a few "Data*" examples there too.

The bottom line is that names are ultimately just random handles and it really doesn't matter that much. But unfortunately they are necessary for making a new repo, so we need to resolve this.

**Reply by rcoreilly** [reply]
At some point we should actually check out the features present in these other frameworks, and steal the good ideas!  https://lakefs.io/blog/jupyter-notebook-10-alternatives-2023/  also there is a lot of bad reactions to jupyter so it would be good to read up on that -- it is obviously very successful and powerful but also various other issues -- they have a "lab" version that is supposed to fix many of the issues.

**Reply by rcoreilly** [reply]
"Lab" is also a possible naming theme, in the overall science space -- like a little laboratory for running experiments with your data or something like that. Notebook is also a central concept / metaphor. Scope-wise however this overall repository is more of a full framework / toolkit and a Notebook would end up being one component of the larger space, which includes a browser and other components, which might include a spreadsheet like thing and who knows what else. Again at some point we probably need to get a bit more specific about what else we want to include so we at least have some kind of overall space and architecture in mind.

**Reply by kkoreilly** [reply]
I like Cogent Lab better than Cogent Vision and Cogent Science, but I am not sure yet if it is good enough.

**Reply by rcoreilly** [reply]
"Cogent Lab" is probably too similar to "CoLab"?

If you look at all the other names in this space, and you think about how it will look on github and as a directory, the individual name (without the cogent prefix) really needs to stand on its own as a "proper name". Just actually picture checking out a repository called `data`, and seeing that name on the list of repositories. It is just too generic. It is like calling your child `child`. It needs to be a _proper_ name, at the top level.

Again, if it is one of many within a whole cluster of cogent apps, like Cogent Mail etc, then this consideration does not apply -- those are automatically "subordinate" within the cogent branding sphere. But this needs to be a "top level" name.

Furthermore, the "cogent" branding will often be dropped, so you want it to be able to stand on its own. Just saying "Data" is not going to cut it -- might as well just call it "software" or "framework".

The very fact that Vision is a metaphorical, non-literal term makes it a proper name. Indeed, it is the proper name of a Marvel character.

Compare what happens when you go to https://en.wikipedia.org and type in Vision vs. Data in their search bar. https://en.wikipedia.org/wiki/Vision has _many_ uses as a proper name for all kinds of things. https://en.wikipedia.org/wiki/Data is just the actual page on data itself as a concept. Yes there are a few other uses of it as a proper name, including a Star Trek character most notably: https://en.wikipedia.org/wiki/Data_(disambiguation)  but it is obvious which name is more generally compelling as a proper name.

**Reply by rcoreilly** [reply]
After further discussion, it seems like `Lab` is actually a good, increasingly widely used "generic" term for this category of software, and there are enough other instances that the similarity with CoLab should not be a problem.

So we'll go for `Cogent Lab`!

**Reply by kkoreilly** [reply]
(For example: MATLAB, JupyterLab, CoLab)
